T21347 update status column on table test

alexandrasp opened the pull request:
Avoid showing unknown on status column while loading test results and
while it is loading show spinning animations waiting for data to arrive.
The data used to show status was also updated to show test count for
each combation of plan and group.

Signed-off-by: Alexandra Pereira <alexandra.pereira@collabora.com>

Commit:
update status column on table test

Avoid showing unknown on status column while loading test results and
while it is loading show spinning animations waiting for data to arrive.
The data used to show status was also updated to show test count for
each combation of plan and group.

Signed-off-by: Alexandra Pereira <alexandra.pereira@collabora.com>

alexandrasp commented:
@gctucker I decided to update to third option(spinning animations with test count). Let's check how it goes on staging. Locally it works, but seems to a bit slow. I also have two doubt for this PR.
First: should spinning animations has a link? 
Second: It is couting all test, is it right ? Or should be just latest?

gctucker commented:
> @gctucker I decided to update to third option(spinning animations with test count). Let's check how it goes on staging. Locally it works, but seems to a bit slow. I also have two doubt for this PR.

It can be slow if the queries don't use any existing index, we can check this later.  I thought the queries would be the same as in other views, but if it's really a new kind and a new index is needed it's easy to add as well.

> First: should spinning animations has a link?

Yes, just like on the other tests and jobs views.

> Second: It is couting all test, is it right ? Or should be just latest?

Each row shows the test cases for a test plan for a kernel revision, so there's no question about "latest" or not.  It should be the same numbers as on the job/branch/kernel view e.g.:

https://staging.kernelci.org/test/job/kernelci/branch/staging.kernelci.org/kernel/staging-20200524.2/

gctucker commented:
Also maybe move the "date" column to the right like on the jobs view, to make it more consistent.

Commit:
update spinning animations

fix pagination so when you change a page
spinning gets updated instantly without any
new request.

Signed-off-by: Alexandra Pereira <alexandra.pereira@collabora.com>

Commit:
add link on spinning animations

create a link on spinning animations
to /test/job/branch/kernel/plan.

Signed-off-by: Alexandra Pereira <alexandra.pereira@collabora.com>

gctucker commented:
@alexandrasp Please rebase when you have a new version, this PR doesn't apply on top of the master branch any more.

Commit:
update execution flow to update table

add an improvement in execution flow to get
tests count and update spinning animations after build
table, making the process quicker than before and using
the same aproach that is already used in job's table.

Signed-off-by: Alexandra Pereira <alexandra.pereira@collabora.com>

gctucker commented:
OK thanks, this seems to be much better.  The only issue left is that it still takes a very long time to load the first time, it took over 12s just to complete the batch query when I tried.  Subsequent updates using the cache take around 1s which is acceptable.

This was already an issue before, when showing "unknown" icons while data was loading.  At least, now it appears more clearly to the user that data is being retrieved, and the numbers are more useful than just a single pass/fail icon.  So this PR is improving things a lot.

I would say it's all fine from a functional point of view.  Looking at the code now.

gctucker commented on app/dashboard/static/js/app/view-tests-all.2020.5.js:
```diff
@@ -43,62 +43,147 @@ require([
-                'plan': plan,
+                'plan': result.name,
             });
 
-            /* Number of test case regressions */
+            // Get total tests count.
+            opId = 'test-total-count-';
+            opId += opIdTail;
             batchOps.push({
                 method: 'GET',
-                operation_id: 'status-' + result._id.$oid,
+                operation_id: opId,
+                resource: 'count',
+                document: 'test_case',
+                query: qStr
+            });
+
+            // Get successful tests count.
+            opId = 'test-success-count-';
+            opId += opIdTail;
+            qHead = 'status=PASS&';
+            qHead += qStr;
+            batchOps.push({
+                method: 'GET',
+                operation_id: opId,
+                resource: 'count',
+                document: 'test_case',
+                query: qHead
+            });
+
+            // Get regressions count.
+            opId = 'test-fail-count-';
+            opId += opIdTail;
+            batchOps.push({
+                method: 'GET',
+                operation_id: opId,
                 resource: 'count',
                 document: 'test_regression',
-                query: qStr,
+                query: qStr
```
I actually think it's better to keep the trailing comma, because if an argument gets added later this line doesn't need to be edited to add a comma.

JSON syntax requires the last item to not have a trailing comma, but this is a Javascript dictionary.

So it's OK either way, but not worth removing the comma imho.

gctucker commented on app/dashboard/static/js/app/view-tests-all.2020.5.js:
```diff
@@ -43,62 +43,147 @@ require([
+            if (gTestCount.hasOwnProperty(opId)) {
+                delete gTestCount[opId];
+            }
+        } else {
+            // Store it in a dictionary for later access.
+            if (!gTestCount.hasOwnProperty(opId)) {
+                gTestCount[opId] = count;
+            }
+        }
```
Are you sure this actually gets used?

When I loaded the page the first time, it took 12s to complete the batch query and there was no test counts shown until the query completed.  My understanding is that it should be showing the ones that can be drawn when it receives them, and store the other ones for later when the user goes to another page.

gctucker replied:
Actually I think the logic here is not for showing results earlier, but to cope with the fact that some elements don't exist when they are on other pages.  So when the data arrives, only the rows that have been drawn can be updated.  The others get updated when the user goes to other pages on the table and other rows get drawn.

This is necessary but unrelated to the performance issue.

gctucker reviewed (commented):
Some minor thing about the coding style.

While I see what you're doing, it's similar to the "jobs" page by showing some results early and storing others that are not in the visible part of the table, I'm not entirely sure it's actually working as expected.  It still seems to be waiting for the batch request to complete before showing anything in the table.

The "jobs" page is a bit slower the first time it loads, but I believe it's still faster than the "tests" page from this PR.  Could you please double check what is going on by debugging the code?

gctucker commented:
I'll check if there are any errors in the logs and if not then I think this PR will be ready to be merged.  The performance is not worse than what is currently in production, but it's better because it's showing progress to the users and test counts which is more useful.

gctucker reviewed (approved):
Tested OK on staging.
